File: lib/cfr/builtins/libs/dart/ui/imageDescriptor.dart
import 'dart:async';
import 'dart:core';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedImageDescriptor extends VMManagedBox<ImageDescriptor> {
  VMManagedImageDescriptor(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['getWidth'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.width,
      ];
    });
    table['getHeight'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.height,
      ];
    });
    table['getBytesPerPixel'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.bytesPerPixel,
      ];
    });
    table['dispose'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.dispose();
      return [];
    });
    table['instantiateCodec'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Future>(
            object: vmObject.instantiateCodec(
                targetHeight: luaCallerArguments.length >= 2
                    ? luaCallerArguments[1]['targetHeight']
                    : null,
                targetWidth: luaCallerArguments.length >= 2
                    ? luaCallerArguments[1]['targetWidth']
                    : null),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ImageDescriptor vmObject;
}

void loadImageDescriptor(
    {required HydroState hydroState, required HydroTable table}) {
  table['imageDescriptorRaw'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<ImageDescriptor>(
          object: ImageDescriptor.raw(
              maybeUnBoxAndBuildArgument<ImmutableBuffer, dynamic>(
                  luaCallerArguments[1],
                  parentState: hydroState),
              rowBytes: luaCallerArguments.length >= 3
                  ? luaCallerArguments[2]['rowBytes']
                  : null,
              height: luaCallerArguments.length >= 3
                  ? luaCallerArguments[2]['height']
                  : null,
              pixelFormat: maybeUnBoxEnum(
                  values: PixelFormat.values,
                  boxedEnum: luaCallerArguments.length >= 3
                      ? luaCallerArguments[2]['pixelFormat']
                      : null),
              width: luaCallerArguments.length >= 3
                  ? luaCallerArguments[2]['width']
                  : null),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['imageDescriptorEncoded'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Future>(
          object: ImageDescriptor.encoded(
              maybeUnBoxAndBuildArgument<ImmutableBuffer, dynamic>(
                  luaCallerArguments[1],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<ImageDescriptor>(boxer: (
      {required ImageDescriptor vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedImageDescriptor(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/imageDescriptor.ts
import { IFuture } from "../async/future";
import { ICodec } from "./codec";
import { IImmutableBuffer } from "./immutableBuffer";
import { PixelFormat } from "./pixelFormat";
declare const dart: {
    ui: {
        imageDescriptorRaw: (
            buffer: IImmutableBuffer,
            props: {
                rowBytes?: number | undefined;
                height: number;
                pixelFormat: PixelFormat;
                width: number;
            }
        ) => IImageDescriptor;
        imageDescriptorEncoded: (
            buffer: IImmutableBuffer
        ) => IFuture<IImageDescriptor>;
    };
};
export interface IImageDescriptor {
    getWidth: () => number;
    getHeight: () => number;
    getBytesPerPixel: () => number;
    dispose: () => void;
    instantiateCodec: (props?: {
        targetHeight?: number | undefined;
        targetWidth?: number | undefined;
    }) => IFuture<ICodec>;
}
export class ImageDescriptor {
    public static raw(
        buffer: IImmutableBuffer,
        props: {
            rowBytes?: number | undefined;
            height: number;
            pixelFormat: PixelFormat;
            width: number;
        }
    ): IImageDescriptor {
        return dart.ui.imageDescriptorRaw(buffer, props);
    }
    public static encoded(buffer: IImmutableBuffer): IFuture<IImageDescriptor> {
        return dart.ui.imageDescriptorEncoded(buffer);
    }
    private readonly _dart_getWidth: () => number = undefined as any;
    private readonly _dart_getHeight: () => number = undefined as any;
    private readonly _dart_getBytesPerPixel: () => number = undefined as any;
    private readonly _dart_dispose: () => void = undefined as any;
    private readonly _dart_instantiateCodec: (props?: {
        targetHeight?: number | undefined;
        targetWidth?: number | undefined;
    }) => IFuture<ICodec> = undefined as any;
    public getWidth(): number {
        return this._dart_getWidth();
    }
    public getHeight(): number {
        return this._dart_getHeight();
    }
    public getBytesPerPixel(): number {
        return this._dart_getBytesPerPixel();
    }
    public dispose(): void {
        return this._dart_dispose();
    }
    public instantiateCodec(props?: {
        targetHeight?: number | undefined;
        targetWidth?: number | undefined;
    }): IFuture<ICodec> {
        return this._dart_instantiateCodec(props);
    }
}
